### PR TITLE
update x509 functions and structures to support using heap hint

### PR DESCRIPTION
**Pull request by jpbland1**
# Description

Adds ex versions of the x509 functions that call XMALLOC or call it by proxy to allow the use of a custom heap.

Fixes zd#17258

# Testing

Updated the api and wolfcrypt test to use the ex versions instead of the original

# Checklist

 - [ ] added tests
 - [ ] updated/added doxygen
 - [ ] updated appropriate READMEs
 - [ ] Updated manual and documentation

### CODE REVIEW

File: src/ssl.c [hunk 1]
```diff
@@ -35201,8 +35220,8 @@ WOLFSSL_X509_PKCS12 *wolfSSL_d2i_PKCS12_fp(XFILE fp,
         WOLFSSL_X509_PKCS12 **pkcs12)
 {
     WOLFSSL_ENTER("wolfSSL_d2i_PKCS12_fp");
-    return (WOLFSSL_X509_PKCS12 *)wolfSSL_d2i_X509_fp_ex(fp, (void **)pkcs12,
-        PKCS12_TYPE);
+    return (WOLFSSL_X509_PKCS12 *)wolfSSL_d2i_X509_fp_ex2(fp, (void **)pkcs12,
```
Comment (dgarske) [inline src/ssl.c:35223]: Please adjust `wolfSSL_d2i_X509_fp_ex` and use directly. Its a static local.

File: wolfssl/openssl/asn1.h [hunk 2]
```diff
@@ -26,16 +26,18 @@
...
-#define ASN1_STRING_type      wolfSSL_ASN1_STRING_type
-#define ASN1_STRING_set       wolfSSL_ASN1_STRING_set
-#define ASN1_OCTET_STRING_set wolfSSL_ASN1_STRING_set
-#define ASN1_STRING_free      wolfSSL_ASN1_STRING_free
-
-#define ASN1_get_object       wolfSSL_ASN1_get_object
-#define d2i_ASN1_OBJECT       wolfSSL_d2i_ASN1_OBJECT
-#define c2i_ASN1_OBJECT       wolfSSL_c2i_ASN1_OBJECT
+#define ASN1_STRING_new           wolfSSL_ASN1_STRING_new
+#define ASN1_STRING_new_ex        wolfSSL_ASN1_STRING_new_ex
```
Comment (dgarske) [inline wolfssl/openssl/asn1.h:30]: Does openssl have `ASN1_STRING_new_ex`? I don't think these macros are needed.

**Review by dgarske — CHANGES_REQUESTED** [review]
Have you tested with `WOLFSSL_HEAP_TEST`?
Do all these new _ex functions need to be public? Can you check code size increase from this change?

**Comment by jpbland1** [comment]
Here's the before and after for the size of the dynamic library with --enable-all:
On commit 5631bc9b2:
```
$ size -G /usr/local/lib/libwolfssl.so.42.0.0 
      text       data        bss      total filename
   1396282     772056    1288960    3457298 /usr/local/lib/libwolfssl.so.42.0.0
```
On latest of this branch:
```
$ size -G /usr/local/lib/libwolfssl.so.42.0.0 
      text       data        bss      total filename
   1398410     776322    1288960    3463692 /usr/local/lib/libwolfssl.so.42.0.0
```
So these changes add about 6.3k, might not be worth it if customers don't need the heap hint for these functions but I didn't want to do a half measure and have to come back and add more functions when a customer does ask for them.

As for testing with `WOLFSSL_HEAP_TEST`, it seems to be broken and doesn't build unless I configure with 
````
./configure --enable-crl --disable-ecc CFLAGS='-DWOLFSSL_HEAP_TEST'
```
And it passes make check.

**Comment by jpbland1** [comment]
please retest this

**Comment by jpbland1** [comment]
retest this please

**Comment by dgarske** [comment]
> Here's the before and after for the size of the dynamic library with --enable-all: On commit [5631bc9](https://github.com/wolfSSL/wolfssl/commit/5631bc9b2d4104aafa34bc956462edea52281812):
> 
> ```
> $ size -G /usr/local/lib/libwolfssl.so.42.0.0 
>       text       data        bss      total filename
>    1396282     772056    1288960    3457298 /usr/local/lib/libwolfssl.so.42.0.0
> ```
> 
> On latest of this branch:
> 
> ```
> $ size -G /usr/local/lib/libwolfssl.so.42.0.0 
>       text       data        bss      total filename
>    1398410     776322    1288960    3463692 /usr/local/lib/libwolfssl.so.42.0.0
> ```
> 
> So these changes add about 6.3k, might not be worth it if customers don't need the heap hint for these functions but I didn't want to do a half measure and have to come back and add more functions when a customer does ask for them.
> 
> As for testing with `WOLFSSL_HEAP_TEST`, it seems to be broken and doesn't build unless I configure with
> 
> ```
> ./configure --enable-crl --disable-ecc CFLAGS='-DWOLFSSL_HEAP_TEST'
> ```
> And it passes make check.
> ```

I think my biggest concern is our `--enable-opensslextra=x509small` or `OPENSSL_EXTRA_X509_SMALL`. We've got a few customers that cannot tolerate code size growth with this.
@JacobBarthelmeh thoughts on this PR and possible code size increase and how to mitigate it?

**Comment by jpbland1** [comment]
It looks like a dtls test failed, having trouble reproducing, lets see if it's just the test
retest this please

**Comment by jpbland1** [comment]
With ./configure --enable-all --enable-opensslextra=x509small
On commit 5631bc9:
```
$ size -G /usr/local/lib/libwolfssl.so.42.0.0 
      text       data        bss      total filename
   1396282     772120    1288960    3457362 /usr/local/lib/libwolfssl.so.42.0.0
```
On latest of this branch:
```
$ size -G /usr/local/lib/libwolfssl.so.42.0.0 
      text       data        bss      total filename
   1398410     776386    1288960    3463756 /usr/local/lib/libwolfssl.so.42.0.0
```
So it's still a 6.3k change.

**Review by JacobBarthelmeh — COMMENTED** [review]
This PR would solve for a specific use case. Would it make sense to add a catch all heap hint? Something along the lines of wolfSSL_SetDefaultHeapHint(void* heap). For cases where a heap hint tied to a specific WOLFSSL_CTX is not a concern and a generic memory pool is registered for catching incoming XMALLOC's with NULL heap hint pointers. Then on XFREE if a generic heap hint has been registered and the heap hint passed to XFREE is NULL it would compare the incoming pointer with the memory buckets in the registered heap hint to free it back to the memory pool. Was thinking of this while reviewing this PR. It would be nice to solve all the potential compatibility layer cases at once.

**Pull request closed without merging**